server/src/future.rs

```diff
@@ -237,27 +237,33 @@ impl Future for ShutdownWaiter {
 		// might never resolve.
 		match Weak::strong_count(&self.0) {
 			0 => Poll::Ready(()),
-			_ => Poll::Pending,
+			n => {
+				tracing::trace!("Shutdown count: {}", n);
+				Poll::Pending
+			}
 		}
 	}
 }
 
 /// A `Future` that resolves once the server has been requested to be stopped.
 #[derive(Debug)]
-pub(crate) struct ServerStopped(Weak<MonitorInner>);
+pub(crate) struct ShutdownRequestedWaiter(Weak<MonitorInner>);
 
-impl Future for ServerStopped {
+impl Future for ShutdownRequestedWaiter {
 	type Output = ();
 
-	fn poll(self: Pin<&mut Self>, _cx: &mut Context) -> Poll<Self::Output> {
+	fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
 		if let Some(arc) = Weak::upgrade(&self.0) {
+			// TODO(niklasad1): this doesn't work.
+			//arc.waker.register(cx.waker());
```

Comment:
Yeah, I guess that waker is only woken when the shutdown finishes, not when it is first requested. You'd need a second waker I think (or to wake that one as soon as shutdown is requested too), but as you said on element, maybe an spmc or something that you can `.close()`/drop and poll for being closed would be less error prone *shrug*.

Reply:
I'll rewrite this with a scmp it won't work anyway for graceful shutdowns for the ws server because the background it's spawned from the hyper service fn and the shutdown doesn't wait for that future anyway I realized.